trace-normalization/src/normalize_utils.rs

```diff
@@ -0,0 +1,205 @@
+// }
+// TODO: implement tag normalization
+// let s: String = normalize_tag(svc);
+// if s == "" {
+//     return (fallbackService(lang), errors::NormalizeErrors::ErrorInvalid)
+// }
+// return (s, err)
+// (svc, None)
+// }
+
+// normalize_name normalizes a span name and returns an error describing the reason
+// (if any) why the name was modified.
+// pub fn normalize_name(name: String) -> (String, Option<errors::NormalizeErrors>) {
+pub fn normalize_name(name: String) -> anyhow::Result<String> {
+    anyhow::ensure!(!name.is_empty(), "Normalizer Error: Empty");
+
+    let mut truncated_name = name.clone();
+
+    if name.len() > MAX_NAME_LEN {
+        truncated_name = truncate_utf8(name, MAX_NAME_LEN);
+    }
+
+    let normalized_name = normalize_metric_names(truncated_name)?;
+    Ok(normalized_name)
+}
+
+// TODO: Implement this in a future PR
+// NormalizeTag applies some normalization to ensure the tags match the backend requirements.
+// pub fn normalize_tag(v: String) -> String {
+// Fast path: Check if the tag is valid and only contains ASCII characters,
+// if yes return it as-is right away. For most use-cases this reduces CPU usage.
+// 	if is_normalized_ascii_tag(v.clone()) {
+// 		return v;
+// 	}
+
+//     if v.is_empty() {
+//         return "".to_string();
+//     }
+
+//     "".to_string()
+// }
+
+// pub fn is_normalized_ascii_tag(tag: String) -> bool {
+//     if tag.is_empty() {
+//         return true;
+//     }
+//     if tag.len() > MAX_TAG_LEN {
+//         return false;
+//     }
+//     if !is_valid_ascii_start_char(tag.chars().next().unwrap()) {
+//         return false;
+//     }
+//     for mut i in 0..tag.len() {
+//         let b: char = tag.chars().nth(i).unwrap();
+//         if is_valid_ascii_tag_char(b) {
+//             continue;
+//         }
+//         if b == '_' {
+//             // an underscore is only okay if followed by a valid non-underscore character
+// 			i+=1;
+// 			if i == tag.len() || !is_valid_ascii_tag_char(tag.chars().nth(i).unwrap()) {
+// 				return false;
+// 			}
+//         } else {
+//             return false;
+//         }
+//     }
+//     true
+// }
+
+// pub fn is_valid_ascii_start_char(c: char) -> bool {
+//     ('a'..='z').contains(&c) || c == ':'
+// }
+
+// pub fn is_valid_ascii_tag_char(c: char) -> bool {
+//     is_valid_ascii_start_char(c) || ('0'..='9').contains(&c) || c == '.' || c == '/' || c == '-'
+// }
+
+pub fn normalize_metric_names(name: String) -> anyhow::Result<String> {
+    anyhow::ensure!(!name.is_empty(), "Normalizer Error: Empty");
+
+    // rust efficient ways to build strings, see here:
+    // https://github.com/hoodie/concatenation_benchmarks-rs
+    let mut result = String::with_capacity(name.len());
+
+    // given a dummy value
+    let mut last_char: char = 'a';
+
+    let char_vec: Vec<char> = name.chars().collect();
+
+    let mut i = 0;
+
+    // skip non-alphabetic characters
+    while i < name.len() && !is_alpha(char_vec[0]) {
```

Comment:
The `0` here seems suspect; is it supposed to be `i` in some form? There's probably a more Rust idiomatic way to express what you are doing, if so.

Reply:
oops, fixed.